src/signature.rs

```diff
@@ -238,15 +239,21 @@ impl <'a> SignedRequest <'a> {
     }
 }
 
+fn hmac<D: Digest>(d: D, key: &[u8], data: &[u8]) -> Vec<u8> {
+    let mut hmac = Hmac::new(d, key);
+    hmac.input(data);
+    hmac.result().code().iter().map(|b| *b).collect::<Vec<u8>>()
+}
+
 fn signature(string_to_sign: &str, signing_key: Vec<u8>) -> String {
-    hmac(SHA256, &signing_key, string_to_sign.as_bytes()).to_hex().to_string()
+    hmac(Sha256::new(), &signing_key, string_to_sign.as_bytes()).to_hex().to_string()
 }
 
 fn signing_key(secret: &str, date: Tm, region: &str, service: &str) -> Vec<u8> {
-    let k_date = hmac(SHA256, format!("AWS4{}", secret).as_bytes(), date.strftime("%Y%m%d").unwrap().to_string().as_bytes());
-    let k_region = hmac(SHA256, &k_date, region.as_bytes());
-    let k_service = hmac(SHA256, &k_region, service.as_bytes());
-    hmac(SHA256, &k_service, b"aws4_request")
+    let k_date = hmac(Sha256::new(), format!("AWS4{}", secret).as_bytes(), date.strftime("%Y%m%d").unwrap().to_string().as_bytes());
+    let k_region = hmac(Sha256::new(), &k_date, region.as_bytes());
+    let k_service = hmac(Sha256::new(), &k_region, service.as_bytes());
+    hmac(Sha256::new(), &k_service, b"aws4_request")
```

Comment:
Oh, no, I was mistaken. The `Hmac::new` takes ownership of the `Digest`:
https://github.com/DaGenix/rust-crypto/blob/e8a513310f5d8e7bf445723bd0231002f144490c/src/hmac.rs#L70